### ISSUE TITLE
WPT errors shouldn't be fatal

### ISSUE THREAD

**Issue by fantasai** [tabatkins/bikeshed#1416]
Making broken or missing WPT tests a fatal error means the spec stops building due to changes in the WPT repository, which isn't acceptable.

(It also means I have to spend a bunch of time stripping out the WPT annotations of specs which have incomplete annotations in order to publish, which is a waste of time.)

**Comment by koto** [comment]
+1, especially given that the test list is hardcoded in the file in bikeshed (which is outdated compared to the current WPT list by a couple of months). This is especially cumbersome for new specs & implementations that actively work on the test files.

**Comment by tabatkins** [comment]
It's absolutely not hardcoded, it's picked up as part of the update process like the definition files.

**Comment by koto** [comment]
Oh, I totally missed that `bikeshed update` exists, sorry!

**Comment by tabatkins** [comment]
Heh, no problem. I need to add a staleness check that complains at you if the data files are older than a week or two.

**Comment by fantasai** [comment]
OK, but can we still get this fixed? Even with up-to-date indexes, having changes in the WPT repo break spec builds is still a problem, and I don't think it makes sense for these errors to be fatal.

**Comment by tabatkins** [comment]
That's very much intentional, tho. If you're using WPT, keeping the spec and the tests healthy and aligned is very important. It's not something you should be ignoring.

**Comment by tabatkins** [comment]
I'm not sure why you think you need to immediately do anything. That's what the `-f` flag is for, after all. And fatal errors don't stop the csswg server from building specs, either.

**Comment by mkruisselbrink** [comment]
FWIW I completely agree here that it seems insane that the only way to work around this is by ignoring all fatal errors. This really doesn't seem like it warrants being treated more strictly than any other link errors.

**Comment by tabatkins** [comment]
I still think link errors *should* be fatal, honestly; they harm the readability of the spec. The reason I relented on that issue, and am resisting on this one, is because of (a) where failures can come from, and (b) how the spec reads when failures happen.

For linking errors, failures can come from *anywhere*. A failure caused by the target definition changing or being removed is important to fix, of course, but failures can also come from *new definitions being added somewhere else* that just happen to make your autolink ambiguous now when it wasn't before. On the other hand, WPT errors can happen for precisely one reason: one of the tests for your spec was moved/renamed/deleted in the WPT repo. Easy to find, easy to fix, and generally rare; plus, as an editor you should be aware of changes to your spec's test folder.

And then when a linking error happens, the spec is still left with an unlinked english version of the term. Based on the linking text, and the context, one can usually still figure out where to look for the definition; this generally just puts us back in the pre-Bikeshed world where cross-spec linking was rare and annoying. When a WPT error happens, tho, we just have... a broken link. Unless you can manually head to the WPT repo and guess what the test path was changed to, it's not easy to figure out what was intended. This is a more severe failure. Even if you're not *displaying* the WPT information in the built spec, it still harms editors' ability to use the inline WPT information if it's wrong, and the longer it remains broken the more difficult it will be to track down the necessary fixes.

Note that this split is consistent with the behavior of failing **biblio** links, which are still fatal errors, and for the same reasons: biblio links shouldn't fail often; when they do, it's something you can fix in one spot in SpecRef; and when they fail, it can be hard to track down what was intended, making it more important to get them working again.

In other words, a failing WPT line harms the usability of the spec in a significant manner, which is my criteria for classifying an error as "fatal". Yes, "significant" is a value judgement.

--------

> I'm not sure why you think standard procedure should be to use -f flags, rather than making these non-fatal? Again, I don't think this morally any worse than link errors, and link errors are nonfatal.

What spec are you working on that is continually having WPT errors regularly enough to make having to use `-f` a "standard procedure"? That sounds like a failure of WPT process that we should address at the source, rather than papering over the problem by relaxing the error classification in Bikeshed.

**Comment by frivoal** [comment]
> WPT errors can happen for precisely one reason: one of the tests for your spec was moved/renamed/deleted in the WPT repo

No, it also happens when a new test is *added* to WPT.

Separately from the rest of the discussion (on which we may or may not agree eventually), can you agree that this at least shouldn't be fatal? Sure, you've asked for a wpt folder to be watched, and being warned that something new was added to it is good. But until you do link to it from the spec, nothing that is in the spec has become incorrect. So by the logic you just explained, that shouldn't be fatal.

**Comment by tabatkins** [comment]
Ah, yes, I'm fine with an unlisted test just being a Warning.

**Comment by fantasai** [comment]
Thanks, @frivoal

**Comment by fantasai** [comment]
I don't mind being forced to fix broken links; broken links should be fixed. It's being forced to triage the entire test suite (and also take on full responsibility for ongoing triage of incoming tests) just to get it to build at all as soon as anyone adds even a single annotation that I have a problem with.

**Comment by tabatkins** [comment]
Yup, switched missing tests to a warning. Your original issue wasn't specific to just that, so I didn't consider it as a separate thing from broken tests. ^_^
